### ISSUE TITLE
Better modularisation?

### ISSUE THREAD

**Issue by alias1** [erocarrera/pefile#124]
First off, love the project and the work that's gone into it, SO useful.

I've been working on a project recently where a binary file (dll) was read as UTF8, and so invalid bytes were corrupted with the utf-8 replacement character. As you could imagine, this makes it rather hard to parse.

I've been slowly making progress with pefile and a ton of reference material, but it's highlighted a point to me that I thought I would raise.

As it currently is, pefile is REALLY good at 'point and click' for a valid binary file, but nowhere near as simple to use piecemeal as an analysis toolkit. Since it will basically error out at the first problem it finds, it never really gives you a chance to properly use what it's already parsed/manually correct things before trying to parse further.

So I decided maybe I could just call each relevant section directly, and build up my own 'parse'. In doing so, I found that the `PE` class has a lot of interwoven dependencies that make it quite hard to pick out the individual pieces needed (without copy/pasting and hacking at it) to parse individual sections. Using `Structure` directly makes this somewhat easier, but still less than ideal.

What I think would end up being awesome, is modularising the relevant parsing sections (including the little checks/fixes along the way), and making them callable without needing an instance of `PE`. That way the main `__parse__` would just end up with a bunch of function calls like `parse_dos_header`, `check_dos_header`, `parse_nt_header`, etc. Each would be a static function that takes all the data it needs as parameters, and returns everything relevant. To maintain the current functionality, the `__parse__` function could then update the relevant vars within the `PE` instance.

This is getting a little long and is a bit mind dumpy, so i'll leave it here, but more than happy to further discuss how I think this would work/help implement it.

**Comment by maximumspatium** [comment]
👍 +1

**Comment by j-t-1** [comment]
This has been a while, so understand if not something you want to continue with.

> As it currently is, pefile is REALLY good at 'point and click' for a valid binary file, but nowhere near as simple to use piecemeal as an analysis toolkit. Since it will basically error out at the first problem it finds, it never really gives you a chance to properly use what it's already parsed/manually correct things before trying to parse further.

Yes, I think pefile is really good at this and encodes a lot of knowledge. The error out situation would probably still hold with the current HEAD.

> So I decided maybe I could just call each relevant section directly, and build up my own 'parse'. In doing so, I found that the PE class has a lot of interwoven dependencies that make it quite hard to pick out the individual pieces needed (without copy/pasting and hacking at it) to parse individual sections. Using Structure directly makes this somewhat easier, but still less than ideal.

If there are small changes to decrease coupling that would be welcome. See also #297.

> What I think would end up being awesome, is modularising the relevant parsing sections (including the little checks/fixes along the way), and making them callable without needing an instance of PE. That way the main __parse__ would just end up with a bunch of function calls like parse_dos_header, check_dos_header, parse_nt_header, etc. Each would be a static function that takes all the data it needs as parameters, and returns everything relevant. To maintain the current functionality, the __parse__ function could then update the relevant vars within the PE instance.

This proposal could be a good idea, although interactions between the components is a complication. As an example, this is mainly relevant to parsing the optional header, but is also relevant to parsing the data-directory entries:
https://github.com/erocarrera/pefile/blob/d508b844e6c775bf9172fe191004b1779867bec2/pefile.py#L3244-L3251

I think the next most important change to pefile is expansion of the test scripts, especially those related to open issues. Many more tests give confidence in making changes, such as those proposed here.
